File: python/kickstarter/model/sunset_air_flight_data.py
from dataclasses import dataclass
from datetime import datetime, timedelta
from decimal import Decimal
import json
from pyflink.common import Types, Row

from model.flight_data import FlightData
from common_functions import serialize

__copyright__  = "Copyright (c) 2024 Jeffrey Jonathan Jennings"
__credits__    = ["Jeffrey Jonathan Jennings"]
__license__    = "MIT"
__maintainer__ = "Jeffrey Jonathan Jennings"
__email__      = "j3@thej3.com"
__status__     = "dev"


@dataclass
class SunsetAirFlightData:
    customer_email_address: str | None
    departure_time: datetime | None
    departure_airport: str | None
    arrival_time: datetime | None
    arrival_airport: str | None
    flight_duration: timedelta | None
    flight_id: str | None
    reference_number: str | None
    total_price: Decimal | None
    aircraft_details: str | None

    def to_flight_data(self):
        return FlightData(email_address=self.customer_email_address,
                          departure_time=self.departure_time,
                          departure_airport_code=self.departure_airport,
                          arrival_time=self.arrival_time,
                          arrival_airport_code=self.arrival_airport,
                          flight_number=self.flight_id,
                          confirmation_code=self.reference_number)
    
    def to_row(self):
        return Row(
            email_address=self.email_address,
            departure_time=serialize(self.flight_departure_time),
            iata_departure_code=self.iata_departure_code,
            arrival_time=serialize(self.flight_arrival_time),
            iata_arrival_code=self.iata_arrival_code,
            flight_number=self.flight_number,
            confirmation=self.confirmation,
            ticket_price=self.ticket_price,
            aircraft=self.aircraft,
            booking_agency_email=self.booking_agency_email
        )

    @staticmethod
    def get_value_type_info():
        return Types.ROW_NAMED(
            field_names=[
                "email_address",
                "flight_departure_time",
                "iata_departure_code",
                "flight_arrival_time",
                "iata_arrival_code",
                "flight_number",
                "confirmation",
                "ticket_price",
                "aircraft",
                "booking_agency_email",
            ],
            field_types=[
                Types.STRING(),
                Types.SQL_TIMESTAMP(),
                Types.STRING(),
                Types.SQL_TIMESTAMP(),
                Types.STRING(),
                Types.STRING(),
                Types.STRING(),
                Types.FLOAT(),
                Types.STRING(),
                Types.STRING(),
            ],
        )
    
    def __eq__(self, other):
        if not isinstance(other, SunsetAirFlightData):
            return False
        return (self.customer_email_address == other.customer_email_address and
                self.departure_time == other.departure_time and
                self.departure_airport == other.departure_airport and
                self.arrival_time == other.arrival_time and
                self.arrival_airport == other.arrival_airport and
                self.flight_duration == other.flight_duration and
                self.flight_id == other.flight_id and
                self.reference_number == other.reference_number and
                self.total_price == other.total_price and
                self.aircraft_details == other.aircraft_details)

    def __hash__(self):
        return hash(self.customer_email_address,
                    self.departure_time,
                    self.departure_airport,
                    self.arrival_time,
                    self.arrival_airport,
                    self.flight_duration,
                    self.flight_id,
                    self.reference_number,
                    self.total_price,
                    self.aircraft_details)

    def __str__(self):
        return json.dumps(self.__dict__, default=str)

File: java/app/src/main/java/kickstarter/model/SunsetAirFlightData.java
/**
 * Copyright (c) 2024 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * 
 */
package kickstarter.model;

import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.*;
import java.math.*;
import java.time.*;
import java.util.*;


@JsonIgnoreProperties(ignoreUnknown = true)
public class SunsetAirFlightData {
    @JsonProperty("email_address")
    private String email_address;

    @JsonProperty("departure_time")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
    private ZonedDateTime departure_time;

    @JsonProperty("departure_airport_code") 
    private String departure_airport_code;

    @JsonProperty("arrival_time")
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
    private ZonedDateTime arrival_time;

    @JsonProperty("arrival_airport_code")
    private String arrival_airport_code;

    @JsonProperty("flight_duration") 
    private Duration flight_duration;

    @JsonProperty("flight_number")
    private String flight_number;

    @JsonProperty("confirmation_code")
    private String confirmation_code;

    @JsonProperty("ticket_price")   
    private BigDecimal ticket_price;

    @JsonProperty("aircraft")
    private String aircraft;

    @JsonProperty("booking_agency_email")
    private String booking_agency_email;

    
    public SunsetAirFlightData() {
        // --- Empty constructor
    }

    public String getEmailAddress() {
        return this.email_address;
    }

    public void setEmailAddress(String EmailAddress) {
        this.email_address = EmailAddress;
    }

    public ZonedDateTime getDepartureTime() {
        return this.departure_time;
    }

    public void setDepartureTime(ZonedDateTime departureTime) {
        this.departure_time = departureTime;
    }

    public String getDepartureAirportCode() {
        return this.departure_airport_code;
    }

    public void setDepartureAirportCode(String departureAirport) {
        this.departure_airport_code = departureAirport;
    }

    public ZonedDateTime getArrivalTime() {
        return this.arrival_time;
    }

    public void setArrivalTime(ZonedDateTime arrivalTime) {
        this.arrival_time = arrivalTime;
    }

    public String getArrivalAirportCode() {
        return this.arrival_airport_code;
    }

    public void setArrivalAirportCode(String arrivalAirport) {
        this.arrival_airport_code = arrivalAirport;
    }

    public Duration getFlightDuration() {
        return this.flight_duration;
    }

    public void setFlightDuration(Duration flightDuration) {
        this.flight_duration = flightDuration;
    }

    public String getFlightNumber() {
        return this.flight_number;
    }

    public void setFlightNumber(String flightId) {
        this.flight_number = flightId;
    }

    public String getConfirmationCode() {
        return this.confirmation_code;
    }

    public void setConfirmationCode(String referenceNumber) {
        this.confirmation_code = referenceNumber;
    }

    public BigDecimal getTicketPrice() {
        return this.ticket_price;
    }

    public void setTicketPrice(BigDecimal totalPrice) {
        this.ticket_price = totalPrice;
    }

    public String getAircraft() {
        return this.aircraft;
    }

    public void setAircraft(String aircraftDetails) {
        this.aircraft = aircraftDetails;
    }

    public String getBookingAgencyEmail() {
        return this.booking_agency_email;
    }

    public void setBookingAgencyEmail(String bookingAgencyEmail) {
        this.booking_agency_email = bookingAgencyEmail;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SunsetAirFlightData that = (SunsetAirFlightData) o;
        return Objects.equals(this.email_address, that.email_address) && 
                                Objects.equals(this.departure_time, that.departure_time) && 
                                Objects.equals(this.departure_airport_code, that.departure_airport_code) && 
                                Objects.equals(this.arrival_time, that.arrival_time) && 
                                Objects.equals(this.arrival_airport_code, that.arrival_airport_code) && 
                                Objects.equals(this.flight_duration, that.flight_duration) && 
                                Objects.equals(this.flight_number, that.flight_number) && 
                                Objects.equals(this.confirmation_code, that.confirmation_code) && 
                                Objects.equals(this.ticket_price, that.ticket_price) && 
                                Objects.equals(this.aircraft, that.aircraft) &&
                                Objects.equals(this.booking_agency_email, that.booking_agency_email);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.email_address,
                            this.departure_time, 
                            this.departure_airport_code, 
                            this.arrival_time, 
                            this.arrival_airport_code, 
                            this.flight_duration, 
                            this.flight_number, 
                            this.confirmation_code, 
                            this.ticket_price, 
                            this.aircraft,
                            this.booking_agency_email);
    }

    @Override
    public String toString() {
        return "SunsetAirFlightData{" +
            "email_address='" + this.email_address + '\'' +
            ", departure_time=" + this.departure_time +
            ", departure_airport_code='" + this.departure_airport_code + '\'' +
            ", arrival_time=" + this.arrival_time +
            ", arrival_airport_code='" + this.arrival_airport_code + '\'' +
            ", flight_duration=" + this.flight_duration +
            ", flight_number='" + this.flight_number + '\'' +
            ", confirmation_code='" + this.confirmation_code + '\'' +
            ", ticket_price=" + this.ticket_price +
            ", aircraft='" + this.aircraft + '\'' +
            ", booking_agency_email='" + booking_agency_email + '\'' +
            '}';
    }

    public FlightData toFlightData() {
        FlightData flightData = new FlightData();

        flightData.setEmailAddress(getEmailAddress());
        flightData.setDepartureTime(getDepartureTime());
        flightData.setDepartureAirportCode(getDepartureAirportCode());
        flightData.setArrivalTime(getArrivalTime());
        flightData.setArrivalAirportCode(getArrivalAirportCode());
        flightData.setFlightNumber(getFlightNumber());
        flightData.setConfirmationCode(getConfirmationCode());

        return flightData;
    }
}
